src/collection-view.js

```diff
@@ -101,12 +101,21 @@ Marionette.CollectionView = Marionette.View.extend({
   _onCollectionAdd: function(child, collection, opts) {
     // `index` is present when adding with `at` since BB 1.2; indexOf fallback for < 1.2
     var index = opts.at !== undefined && (opts.index || collection.indexOf(child));
+    var models = collection.models;
 
     // When filtered or when there is no initial index, calculate index.
-    if (this.getOption('filter') || index === false) {
-      index = _.indexOf(this._filteredSortedModels(index), child);
+    if (this.getOption('filter') || this.getViewComparator() || index === false) {
+      models = this._filteredSortedModels(index);
     }
 
+    // find adjusted index
+    _.reduce(models, function(counter, model) {
+      var view = this.children.findByModel(model);
+      if (view) { return view._index; }
+
+      return model === child ? index = counter + 1 : counter;
+    }, -1, this);
```

Comment:
The logic here feels a little tricky to follow. Can you tell me if I've understood it correctly?

1. You go through until you find the view to insert after
2. On subsequent passes, you just return that index

The reduce call doesn't get assigned to anything so I'm assuming that the index variable set on line 103 is already correct in this case?

The other case here is the view is new, so `view` is always `undefined`, in which case:

1. Compare model to child
2. When you find it, set index to the next item

I'm assuming that this second path is how the view gets added in the right place?